lightly/loss/dcl_loss.py

```diff
@@ -0,0 +1,219 @@
+from typing import Callable, Optional
+from functools import partial
+
+import torch
+from torch import Tensor
+
+from lightly.utils import dist
+
+def negative_mises_fisher_weights(
+    out0: Tensor, 
+    out1: Tensor, 
+    sigma: float=0.5
+) -> torch.Tensor:
+    """Negative Mises-Fisher weighting function as presented in Decoupled
+    Contrastive Learning [0].
+
+    The implementation was inspired by [1].
+    
+    - [0] Chun-Hsiao Y. et. al., 2021, Decoupled Contrastive Learning https://arxiv.org/abs/2110.06848
+    - [1] https://github.com/raminnakhli/Decoupled-Contrastive-Learning
+
+    Args:
+        out0:
+            Output projections of the first set of transformed images.
+            Shape: (batch_size, embedding_size)
+        out1:
+            Output projections of the second set of transformed images.
+            Shape: (batch_size, embedding_size)
+        sigma:
+            Similarities are scaled by inverse sigma.
+    Returns:
+        A tensor with shape (batch_size,) where each entry is the weight for one
+        of the input images.
+    
+    """
+    similarity = torch.einsum('nm,nm->n', out0, out1) / sigma
+    return 2 - out0.shape[0] * torch.nn.functional.softmax(similarity, dim=0)
+
+class DCLLoss(torch.nn.Module):
+    """Implementation of the Decoupled Contrastive Learning Loss from
+    Decoupled Contrastive Learning [0].
+    
+    This code implements Equation 6 in [0], including the sum over all images `i`
+    and views `k`. The loss is reduced to a mean loss over the mini-batch.
+    The implementation was inspired by [1].
+
+    - [0] Chun-Hsiao Y. et. al., 2021, Decoupled Contrastive Learning https://arxiv.org/abs/2110.06848
+    - [1] https://github.com/raminnakhli/Decoupled-Contrastive-Learning
+
+    Attributes:
+        temperature:
+            Similarities are scaled by inverse temperature.
+        weight_fn:
+            Weighting function `w` from the paper. No weighting is performed if
+            weight_fn is None.
+        gather_distributed:
+            If True then negatives from all gpus are gathered before the 
+            loss calculation.
+    
+    Examples:
+
+        >>> loss_fn = DCLLoss(temperature=0.07)
+        >>>
+        >>> # generate two random transforms of images
+        >>> t0 = transforms(images)
+        >>> t1 = transforms(images)
+        >>>
+        >>> # embed images using some model, for example SimCLR
+        >>> out0 = model(t0)
+        >>> out1 = model(t1)
+        >>>
+        >>> # calculate loss
+        >>> loss = loss_fn(out0, out1)
+        
+    """
+    def __init__(
+        self,
+        temperature: float = 0.1,
+        weight_fn: Optional[Callable[[Tensor, Tensor], Tensor]] = None,
```

Comment:
If we add such "complex" type hints I'd expect somewhere examples/ explanations on what this means. Ideally, above where you describe the attribute.